src/Microsoft.AspNetCore.Blazor/Components/BlazorComponent.cs

```diff
@@ -184,5 +201,24 @@ void IHandleEvent.HandleEvent(UIEventHandler handler, UIEventArgs args)
             // at the end of every event callback.
             StateHasChanged();
         }
+
+        void IHandleAfterRender.OnAfterRender()
```

Comment:
We're lacking unit test coverage for the `BlazorComponent` base class presently. Initially that was deliberate because the set of lifecycle methods was constantly in flux, and my goal was to keep the `BlazorComponent` small and simple enough to reason about trivially.

But we're now reaching the end of that phase I think. We should do some kind of design review of the lifecycle methods that now exist (for naming, exact semantics, etc.) and then nail it all down with tests.

The E2E coverage of it is reasonable already though.

Reply:
I'm on board with doing a conceptual review once the features are in. I think what's here follows the same pattern as the other stuff we have and that's good enough for right now.

Reply:
Great. One other thing to include in a design review about this (and other lifecycle methods) is failure handling. Currently if you have an async exception we deal with it, but synchronous ones kill the entire render batch. That’s probably not what we want!